github_app.go

```diff
@@ -302,3 +345,57 @@ func (g *GithubAppAuth) RedactPrivateKey() *GithubAppAuth {
 	g.PrivateKey = []byte(RedactedValue)
 	return g
 }
+
+// cachedInstallationToken represents a GitHub installation token that's
+// cached in memory.
+type cachedInstallationToken struct {
+	installationToken string
+	expiresAt         time.Time
+}
+
+func (c *cachedInstallationToken) isExpired(lifetime time.Duration) bool {
+	return time.Until(c.expiresAt) < lifetime
+}
+
+// installationTokenCache is a concurrency-safe cache mapping the installation
+// ID to the cached GitHub installation token for it.
+type installationTokenCache struct {
+	cache map[int64]cachedInstallationToken
+	mu    sync.RWMutex
+}
+
+// ghInstallationTokenCache is the in-memory instance of the cache for GitHub
+// installation tokens.
+var ghInstallationTokenCache = installationTokenCache{
+	cache: make(map[int64]cachedInstallationToken),
+	mu:    sync.RWMutex{},
+}
+
+// get gets an installation token from the cache by its installation ID. It will
+// not return a token if the token will expire before the requested lifetime.
+func (c *installationTokenCache) get(installationID int64, lifetime time.Duration) string {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	cachedToken, ok := c.cache[installationID]
+	if !ok {
+		return ""
+	}
+	if cachedToken.isExpired(lifetime) {
```

Comment:
Do we have some sort of token limit such that if a bunch of things are long running spawn new tokens and they aren't revoked after nothing is using it we're going to have a new issue?

Reply:
oop i do not know how this got commented twice